File: convert_md_to_docx.py
#!/usr/bin/env python3
"""
Convert Markdown to DOCX
Converts HARDCODED_VALUES_FULL_AUDIT.md to DOCX format
"""

from docx import Document
from docx.shared import Pt, RGBColor, Inches
from docx.enum.text import WD_ALIGN_PARAGRAPH
import re

def read_markdown(filepath):
    """Read markdown file"""
    with open(filepath, 'r', encoding='utf-8') as f:
        return f.read()

def add_heading(doc, text, level=1):
    """Add heading to document"""
    doc.add_heading(text, level=level)

def add_paragraph(doc, text, bold=False, italic=False):
    """Add paragraph to document"""
    p = doc.add_paragraph(text)
    if bold or italic:
        for run in p.runs:
            run.bold = bold
            run.italic = italic
    return p

def parse_table(lines, start_idx):
    """Parse markdown table and return table data and next index"""
    headers = []
    rows = []
    idx = start_idx
    
    # Parse header
    if idx < len(lines) and '|' in lines[idx]:
        header_line = lines[idx].strip()
        headers = [h.strip() for h in header_line.split('|')[1:-1]]
        idx += 1
        
        # Skip separator line
        if idx < len(lines) and '---' in lines[idx]:
            idx += 1
        
        # Parse data rows
        while idx < len(lines) and '|' in lines[idx]:
            row_line = lines[idx].strip()
            if not row_line or row_line.startswith('#'):
                break
            row = [cell.strip() for cell in row_line.split('|')[1:-1]]
            if row and any(cell for cell in row):  # Only add non-empty rows
                rows.append(row)
            idx += 1
    
    return headers, rows, idx

def add_table_to_doc(doc, headers, rows):
    """Add table to document"""
    if not headers or not rows:
        return
    
    table = doc.add_table(rows=1, cols=len(headers))
    table.style = 'Light Grid Accent 1'
    
    # Add headers
    header_cells = table.rows[0].cells
    for i, header in enumerate(headers):
        header_cells[i].text = header
        # Make header bold
        for paragraph in header_cells[i].paragraphs:
            for run in paragraph.runs:
                run.bold = True
    
    # Add rows
    for row in rows:
        row_cells = table.add_row().cells
        for i, cell_text in enumerate(row):
            row_cells[i].text = cell_text

def markdown_to_docx(md_filepath, docx_filepath):
    """Convert markdown to DOCX"""
    doc = Document()
    
    # Add title
    title = doc.add_heading('OneMind AI - Complete Hardcoded Values Audit', 0)
    title.alignment = WD_ALIGN_PARAGRAPH.CENTER
    
    content = read_markdown(md_filepath)
    lines = content.split('\n')
    
    i = 0
    while i < len(lines):
        line = lines[i]
        
        # Skip empty lines
        if not line.strip():
            i += 1
            continue
        
        # Handle headings
        if line.startswith('# '):
            add_heading(doc, line[2:].strip(), 1)
            i += 1
        elif line.startswith('## '):
            add_heading(doc, line[3:].strip(), 2)
            i += 1
        elif line.startswith('### '):
            add_heading(doc, line[4:].strip(), 3)
            i += 1
        elif line.startswith('#### '):
            add_heading(doc, line[5:].strip(), 4)
            i += 1
        
        # Handle tables
        elif '|' in line:
            headers, rows, next_idx = parse_table(lines, i)
            if headers and rows:
                add_table_to_doc(doc, headers, rows)
            i = next_idx
        
        # Handle bold text (markdown style)
        elif line.startswith('**'):
            text = line.replace('**', '')
            add_paragraph(doc, text, bold=True)
            i += 1
        
        # Handle regular paragraphs
        elif line.strip() and not line.startswith('---'):
            # Clean up markdown formatting
            text = line.replace('**', '').replace('`', '').replace('_', '')
            if text.strip():
                add_paragraph(doc, text)
            i += 1
        
        # Handle horizontal rules
        elif line.startswith('---'):
            doc.add_paragraph()
            i += 1
        
        else:
            i += 1
    
    # Save document
    doc.save(docx_filepath)
    print(f"✅ Successfully converted to: {docx_filepath}")

if __name__ == '__main__':
    md_file = r'c:\Projects\OneMindAI\docs\HARDCODED_VALUES_FULL_AUDIT.md'
    docx_file = r'c:\Projects\OneMindAI\docs\HARDCODED_VALUES_FULL_AUDIT.docx'
    
    try:
        markdown_to_docx(md_file, docx_file)
    except Exception as e:
        print(f"❌ Error: {e}")
        import traceback
        traceback.print_exc()

File: convert_md_to_docx.js
#!/usr/bin/env node

const fs = require('fs');
const path = require('path');

// Simple markdown to HTML converter, then we'll create a basic DOCX
const { Document, Packer, Paragraph, Table, TableCell, TableRow, BorderStyle, WidthType, TextRun, HeadingLevel, AlignmentType } = require('docx');

function parseMarkdownFile(filePath) {
  const content = fs.readFileSync(filePath, 'utf-8');
  return content;
}

function createDocxFromMarkdown(mdContent) {
  const lines = mdContent.split('\n');
  const sections = [];
  let currentSection = [];
  
  for (let i = 0; i < lines.length; i++) {
    const line = lines[i];
    
    // Skip empty lines at start of section
    if (currentSection.length === 0 && !line.trim()) continue;
    
    currentSection.push(line);
    
    // Process section when we hit a heading or end of file
    if ((line.startsWith('#') && currentSection.length > 1) || i === lines.length - 1) {
      if (currentSection.length > 1) {
        sections.push(currentSection.slice(0, -1).join('\n'));
      }
      currentSection = [line];
    }
  }
  
  return sections;
}

function convertToDocxParagraphs(mdContent) {
  const lines = mdContent.split('\n');
  const paragraphs = [];
  let i = 0;
  
  while (i < lines.length) {
    const line = lines[i];
    
    // H1 Heading
    if (line.startsWith('# ')) {
      paragraphs.push(
        new Paragraph({
          text: line.substring(2).trim(),
          heading: HeadingLevel.HEADING_1,
          thematicBreak: false,
          spacing: { after: 200 }
        })
      );
    }
    // H2 Heading
    else if (line.startsWith('## ')) {
      paragraphs.push(
        new Paragraph({
          text: line.substring(3).trim(),
          heading: HeadingLevel.HEADING_2,
          spacing: { after: 150 }
        })
      );
    }
    // H3 Heading
    else if (line.startsWith('### ')) {
      paragraphs.push(
        new Paragraph({
          text: line.substring(4).trim(),
          heading: HeadingLevel.HEADING_3,
          spacing: { after: 100 }
        })
      );
    }
    // H4 Heading
    else if (line.startsWith('#### ')) {
      paragraphs.push(
        new Paragraph({
          text: line.substring(5).trim(),
          heading: HeadingLevel.HEADING_4,
          spacing: { after: 100 }
        })
      );
    }
    // Code block
    else if (line.startsWith('```')) {
      const codeLines = [];
      i++;
      while (i < lines.length && !lines[i].startsWith('```')) {
        codeLines.push(lines[i]);
        i++;
      }
      
      paragraphs.push(
        new Paragraph({
          text: codeLines.join('\n'),
          style: 'Intense Quote',
          spacing: { before: 100, after: 100 }
        })
      );
    }
    // Table
    else if (line.trim().startsWith('|')) {
      const tableLines = [];
      while (i < lines.length && lines[i].trim().startsWith('|')) {
        tableLines.push(lines[i]);
        i++;
      }
      i--;
      
      // Parse table
      const rows = tableLines.map(l => 
        l.split('|').slice(1, -1).map(cell => cell.trim())
      ).filter(row => row.some(cell => cell && cell !== '---'));
      
      if (rows.length > 0) {
        const tableRows = rows.map((row, idx) => 
          new TableRow({
            cells: row.map(cellText => 
              new TableCell({
                children: [new Paragraph({
                  text: cellText,
                  bold: idx === 0
                })],
                width: { size: 100 / row.length, type: WidthType.PERCENTAGE }
              })
            )
          })
        );
        
        paragraphs.push(
          new Table({
            rows: tableRows,
            width: { size: 100, type: WidthType.PERCENTAGE }
          })
        );
      }
    }
    // Bullet list
    else if (line.trim().startsWith('- ')) {
      paragraphs.push(
        new Paragraph({
          text: line.trim().substring(2),
          bullet: { level: 0 },
          spacing: { after: 50 }
        })
      );
    }
    // Regular paragraph
    else if (line.trim()) {
      paragraphs.push(
        new Paragraph({
          text: line.trim(),
          spacing: { after: 100 }
        })
      );
    }
    
    i++;
  }
  
  return paragraphs;
}

async function main() {
  try {
    const mdFile = path.join(__dirname, 'docs', 'API_CONFIG_FINAL_AUDIT_REPORT.md');
    const docxFile = path.join(__dirname, 'docs', 'API_CONFIG_FINAL_AUDIT_REPORT.docx');
    
    console.log('📖 Reading markdown file...');
    const mdContent = parseMarkdownFile(mdFile);
    
    console.log('🔄 Converting to DOCX format...');
    const paragraphs = convertToDocxParagraphs(mdContent);
    
    const doc = new Document({
      sections: [{
        properties: {},
        children: paragraphs
      }]
    });
    
    console.log('💾 Writing Word document...');
    const buffer = await Packer.toBuffer(doc);
    fs.writeFileSync(docxFile, buffer);
    
    console.log('✅ Success!');
    console.log(`📄 Word document created: ${docxFile}`);
    console.log(`📊 File size: ${(buffer.length / 1024).toFixed(2)} KB`);
    
  } catch (error) {
    console.error('❌ Error:', error.message);
    process.exit(1);
  }
}

main();
